sphinx/config.py

```diff
@@ -541,6 +556,33 @@ def eval_config_file(filename: str, tags: Tags | None) -> dict[str, Any]:
             msg = __("There is a programmable error in your configuration file:\n\n%s")
             raise ConfigError(msg % traceback.format_exc()) from exc
 
+        try:
+            # Attempt to identify simple constant-value assignments in the conf.py file.
+            # Truly constant variables don't exist in Python, so it is possible that
+            # subsequent code may re-assign to these names.
+            constant_assignments = [
+                node for node in ast.iter_child_nodes(ast.parse(conf_py))
+                if isinstance(node, ast.Assign)
+                and all(isinstance(target, ast.Name) for target in node.targets)
+                and (
+                    isinstance(node.value, ast.Constant)
+                    or (
+                        isinstance(node.value, ast.List | ast.Tuple)
+                        and all(isinstance(elt, ast.Constant) for elt in node.value.elts)
+                    )
+                )
+            ]
+            namespace['__constants__'] = frozenset(chain.from_iterable(
+                (
+                    target.id
+                    for target in name.targets
+                    if hasattr(target, 'id') and target.id in Config.config_values
+                )
+                for name in constant_assignments
+            ))
+        except Exception:
```

Comment:
`tree = ast.parse(code_py)` can never fail since compiled the code above. 
In addition, if someone writes

```
x = 1
x = get_x()
```

and 

```
x = get_x()
x = 1
```

Then the first one is dynamic while the second can be considered constant. Your filter unfortunately doesn't consider those cases. As such, I think this approach is too complex.